ai_economist/foundation/scenarios/simple_wood_and_stone/layout_from_file.py

```diff
@@ -612,3 +657,156 @@ def scenario_metrics(self):
+        elif isinstance(skill_rank_of_top_agents, (tuple, list)):
+            self.skill_rank_of_top_agents = list(set(skill_rank_of_top_agents))
+        else:
+            raise TypeError(
+                "skill_rank_of_top_agents must be a scalar "
+                "index, or a list of scalar indices."
+            )
+        for rank in self.skill_rank_of_top_agents:
+            assert 0 <= rank < self.n_agents
+        assert 0 < len(self.skill_rank_of_top_agents) < self.n_agents
+
+        # Set the skill associated with each skill rank
+        bm = self.get_component("Build")
+        assert bm.skill_dist == "pareto"
+        pmsm = bm.payment_max_skill_multiplier
+        ranked_skills = np.array(
+            [
+                np.sort(
+                    np.minimum(
+                        pmsm, (pmsm - 1) * np.random.pareto(4, size=self.n_agents) + 1
+                    ),
+                )
+                for _ in range(100000)
+            ]
+        )
+        self._avg_ranked_skill = ranked_skills.mean(axis=0) * bm.payment
+        # Reverse the order so index 0 is the highest-skilled
+        self._avg_ranked_skill = self._avg_ranked_skill[::-1]
+
+    def additional_reset_steps(self):
+        """
+        Extra scenario-specific steps that should be performed at the end of the reset
+        cycle.
+
+        For each reset cycle...
+            First, reset_starting_layout() and reset_agent_states() will be called.
+
+            Second, <component>.reset() will be called for each registered component.
+
+            Lastly, this method will be called to allow for any final customization of
+            the reset cycle.
+
+        For this scenario, this method resets optimization metric trackers. This is
+        where each agent gets assigned to one of the skills and the starting
+        locations are reset according to self.skill_rank_of_top_agents.
+        """
+        self.world.clear_agent_locs()
+        for i, agent in enumerate(self.world.get_random_order_agents()):
+            agent.state["build_payment"] = self._avg_ranked_skill[i]
+            if i in self.skill_rank_of_top_agents:
+                r_min, r_max = 0, self._water_line
+            else:
+                r_min, r_max = self._water_line + 1, self.world_size[0]
+
+            r = np.random.randint(r_min, r_max)
+            c = np.random.randint(0, self.world_size[1])
+            n_tries = 0
+            while not self.world.can_agent_occupy(r, c, agent):
+                r = np.random.randint(r_min, r_max)
+                c = np.random.randint(0, self.world_size[1])
+                n_tries += 1
+                if n_tries > 200:
+                    raise TimeoutError
+            self.world.set_agent_loc(agent, r, c)
+
+        # compute current objectives
+        curr_optimization_metric = self.get_current_optimization_metrics()
+
+        self.curr_optimization_metric = deepcopy(curr_optimization_metric)
```

Comment:
this is slow -- consider wrapping this in a numpy array and do fast array creation